File: pRobot169/python/dump_lists.py
import sys
import time
import socket
import gzip
from io import BytesIO

fAppend = sys.argv[1]
hh = sys.argv[2]
document = sys.argv[3:]

if not document:
    sys.exit()

def dump1list(host, fileRelative):
    timeInit = time.time()
    port = 80

    EOL = "\r\n"
    msgGetGzip = (
        f"GET /{fileRelative} HTTP/1.0{EOL}"
        f"Host: {host}{EOL}"
        "Accept: text/html, text/plain, text/sgml, */*;q=0.01{EOL}"
        "Accept-Encoding: gzip, compress{EOL}"
        "Accept-Language: en{EOL}"
        "User-Agent: Python{EOL}"
        "{EOL}"
    )

    msgGet = msgGetGzip

    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.connect((host, port))
    sock.sendall(msgGet.encode() + EOL.encode())

    print("getting...")
    response = b""
    sock.settimeout(0.05)
    while True:
        try:
            buf = sock.recv(10000)
            if not buf:
                break
            response += buf
        except socket.timeout:
            if time.time() - timeInit > 10:
                break

    sock.close()

    idx_rnrn = response.find(b"\r\n\r\n")
    if idx_rnrn < 0:
        return ""

    header = response[:idx_rnrn+4].decode().split("\r\n")
    content = response[idx_rnrn+4:]

    isVaryAcceptEncoding = False
    contentLength = 0
    contentEncoding = "text"
    for line in header:
        if "Vary:" in line and "Accept-Encoding" in line:
            isVaryAcceptEncoding = True
        elif line.startswith("Content-Length:"):
            contentLength = int(line.split(" ")[1])
        elif line.startswith("Content-Encoding:"):
            contentEncoding = line.split(" ")[1].strip()

    if (len(content) != contentLength and contentLength > 0) or len(content) == 0:
        return ""

    if "gzip" in contentEncoding:
        content = gzip.GzipFile(fileobj=BytesIO(content)).read()

    return htm2list(content.decode())

def htm2list(content):
    lineBuf = [""] * 12
    outlist = ""
    for line in content.split("\n"):
        line = line.strip()
        lineBuf.pop(0)
        lineBuf.append(line)

        if "readPost" in lineBuf[9] or ("queryUser" in lineBuf[5] and lineBuf[9].startswith("<font")):
            num = int(lineBuf[1].split(">")[1].split("<")[0])
            mark = lineBuf[3].split("</td>")[0].lower()[0]
            date = lineBuf[4].split("</td>")[0].split("../")[1]
            author = lineBuf[5].split("</a>")[0].split(">")[1]

            if '"javascript:thread(' in lineBuf[7]:
                title = lineBuf[7].split("'")[1]
            elif lineBuf[9].startswith("<font"):
                title = lineBuf[9].split("</font>")[0].split(">")[1]

            title = "◇ " + title if not title.startswith("Re: ") else "Re " + title[4:]
            title = title.replace("&gt;", ">").replace("&lt;", "<").replace("&nbsp;", " ").replace("&quot;", "\"").replace("&amp;", "&")

            gy = lineBuf[10].split("<em>")[1].split("</em>")[0] if "<em>" in lineBuf[10] else ""
            gy = min(int(gy), 99) if gy else ""
            gy = f"{gy:02}" if gy else ""

            outlist += f" {num:5} {mark:1}{gy:2}{date:5} {author:12} {title}\r\n"

    return outlist

with open(fAppend, "ab") as hdmp:
    for doc in document:
        print(f"http://{hh}/{doc}")
        list_content = dump1list(hh, doc)
        if list_content:
            print(" -> get")
            hdmp.write(list_content.encode())
        else:
            print(" -> fail")

File: pRobot169/perl/dump_lists.pl
#!/usr/bin/perl
#
use strict;
use IO::Handle;
my $fAppend = shift || exit;
my $hh = shift || exit;
@ARGV > 0 or exit;

my @document = @ARGV;
sub dump1list;
my $hdmp;
open $hdmp, ">>$fAppend" or die;
binmode $hdmp;
$hdmp->autoflush(1);


foreach my $i (0 .. $#document) {
    my $document = $document[$i];
    my $list;
    $list = "";
    print "http://$hh/$document";
    $list = dump1list($hh, $document);
    if ($list ne "") {
        print " -> get\n";
        print $hdmp $list;
    }
    else {
        print " -> fail\n";
    }
}

close $hdmp;
exit;

sub dump1list ($$) {
    my ($host, $fileRelative) = @_;
    my $timeInit = time;
    my $port = 80;

    # --- try to find if the system has gzip --------
    my $EOL = "\015\012";
    my $msgGetGzip = 
"GET /$fileRelative HTTP/1.0".$EOL.
"Host: $host".$EOL.
"Accept: text/html, text/plain, text/sgml, */*;q=0.01".$EOL.
"Accept-Encoding: gzip, compress".$EOL.
"Accept-Language: en".$EOL.
"User-Agent: Perl".$EOL.
"";

    my $msgGet = $msgGetGzip;
    # ===============================================

    # --- send the request ----------------------------
    use IO::Socket;
    # http://webbbs.gamer.com.tw/board.php?brd=Chat
    my $sock = IO::Socket::INET->new("${host}:${port}") or die "socket create error\n";
    $sock->print($msgGet.$EOL);
    # =================================================

    # --- setup some variables for select(...) ------
    my ($rin, $win, $ein);
    $rin = $win = $ein = '';
    vec($rin, fileno($sock), 1) = 1;
    vec($win, fileno($sock), 1) = 1;
    $ein = $rin | $win;
    # ===============================================

    print "getting...";
    # --- get the response ------------------
    my $response = "";
    while (! $sock->eof) {
        my ($rout, $wout, $eout) = ($rin, $win, $ein);
        my ($nfound, $timeleft) = select($rout, $wout, $eout, 0.05);
        if ($nfound > 0) {
            my $buf;
            $sock->read($buf, 10000);
            $response.=$buf;
        }
        last if (time - $timeInit > 10);
    }
    $sock->close;
    select(undef, undef, undef, 0.05) while (time == $timeInit);
    # =======================================

    # --- parse the response -------------------
    my $idx_rnrn = index($response, "\r\n\r\n");
    return "" if ($idx_rnrn < 0);

    my @header = split(/\r\n/, substr($response, 0, $idx_rnrn+1));
    my $content = substr($response, $idx_rnrn+4);

    my $isVaryAcceptEncoding = 0;
    my $contentLength = 0;
    my $contentEncoding = "text";
    foreach (@header) {
        s/[\r\n]//g;
        if (/^Vary:/ && /Accept-Encoding/) {
            $isVaryAcceptEncoding = 1;
        }
        elsif (/^Content-Length: (\d+)$/) {
            $contentLength = $1 + 0;
        }
        elsif (/^Content-Encoding: (.+)$/) {
            $contentEncoding = $1;
        }
    }
    # ==========================================

    # --- exit if the content is inconsistent ---------
    if ((length($content) != $contentLength && $contentLength > 0)
        || length($content) == 0) {
        # print STDERR "length unmatch\n";
        return "";
    }
    # =================================================

    # --- print the content -------------
    if ($contentEncoding =~ /gzip/) {
        use Compress::Zlib;
        $content = Compress::Zlib::memGunzip($content);
    }
    return &htm2list($content);
    # ===================================
}

exit;

sub htm2list ($) {
#    my $content = shift;
#    my @list = grep(/./,split(/\n/,$content));
    my @lineBuf;
    for my $i (0 .. 11) {
        $lineBuf[$i] = "";
    }
    my ($num, $mark, $gy, $date, $author, $title);

    my $outlist = "";
    for ( grep(/./, split(/\n/, shift)) ) {
        chomp;
        shift @lineBuf;
        push @lineBuf, $_;

        if ($lineBuf[9] =~ /readPost/ || ($lineBuf[5] =~ /queryUser/ && 
                $lineBuf[9] =~ /^<font.*>/)) {
            $lineBuf[1] =~ /^<.*>(\d+)<\/.*>/;
            $num = $1;

            $lineBuf[3] =~ m#^(.*)</td>#;
            $mark = lc(substr($1." ",0,1));

            $lineBuf[4] =~ m#^<td.*>../(../..)</td>#;
            $date = $1;

            $lineBuf[5] =~ m#^<td><a .*>(.*)</a></td>#;
            $author = $1;

            if ($lineBuf[7] =~ /\"javascript:thread\(\'(.*)\'\)\"/) {
                $title = $1;
            } elsif ($lineBuf[9] =~ m#<font.*>(.*)</font></td>#) {
                $title = $1;
            }

            $title = "◇ ".$title if $title !~ /^Re: /;
            $title = "Re ".substr($title,4) if $title =~ /^Re: /;
            $title =~ s/&gt;/>/g;
            $title =~ s/&lt;/</g;
            $title =~ s/&nbsp;/ /g;
            $title =~ s/&quot;/\"/g;
            $title =~ s/&amp;/\&/g;

            if ($lineBuf[10] =~ m#<td><em>(\d+)</em></td>#) {
                $gy = $1;
                $gy = 99 if $gy>99;
                $gy = sprintf("%02d",$gy);
            }
            else {
                $gy = "";
            }

            $outlist .= sprintf(" %5d %1s%2s%5s %-12s %s\015\012",
                $num, $mark, $gy, $date, $author, $title);
        }
    }
    return $outlist;
}